torch/_ops.py

```diff
@@ -925,6 +945,94 @@ def tags(self):
+            DispatchKey.PythonTLSSnapshot,
+            DispatchKey.PythonDispatcher,
+        ]
+
+        def _may_use_fallthrough_instead_of_fallback(key: DispatchKey):
+            if torch._C._dispatch_has_kernel_for_dispatch_key(self.name(), key):
+                return torch._C._dispatch_kernel_for_dispatch_key_is_fallthrough(
+                    self.name(), key
+                )
+
+            return (
+                key not in self.py_kernels
+                or self.py_kernels[key] is torch.library.fallthrough_kernel
+            )
+
+        return [
+            key
+            for key in _DEFAULT_FALLTHROUGH_KEYS
+            if _may_use_fallthrough_instead_of_fallback(key)
+        ]
+
+    def __call__(self, /, *args, **kwargs):
+        return self._dispatch_in_python(args, kwargs, self._fallthrough_keys())
+
+    def redispatch(self, /, keyset, *args, **kwargs):
+        return self._dispatch_in_python(args, kwargs, self._fallthrough_keys())
+
+    def _dispatch_in_python(self, args, kwargs, fallthrough_keys):
+        non_fallthrough_keys = torch._C._dispatch_keyset_full()
+        for key in fallthrough_keys:
+            non_fallthrough_keys = non_fallthrough_keys.remove(key)
+
+        dispatch_key_set = _compute_keyset(args, kwargs, non_fallthrough_keys)
+        dispatch_key = dispatch_key_set.highestPriorityTypeId()
+
+        handler = (
+            self._get_dispatch(dispatch_key)
+            if dispatch_key not in self._dispatch_cache
+            else self._dispatch_cache[dispatch_key]
+        )
+
+        if isinstance(handler, DispatchKey):
+            # fallthrough keys can be registered at runtime via torch.library.impl
+            # so need to add it to fallthrough_keys and re-dispatch.
+            if torch._C._dispatch_kernel_for_dispatch_key_is_fallthrough(
+                self.name(), dispatch_key
+            ):
+                return self._dispatch_in_python(
+                    args, kwargs, fallthrough_keys + [dispatch_key]
+                )
+
+            raise RuntimeError(
+                f"Custom op {self} received a Pytree input when dispatching {handler}."
+                f" but no python implementation is found."
+                f" Please file an issue on this when you encounter this error."
+            )
+
+        assert isinstance(handler, Callable)  # type: ignore[arg-type]
+        if isinstance(handler, Kernel) and handler.with_keyset:
+            return handler(dispatch_key_set, *args, **kwargs)
+        else:
+            return handler(*args, **kwargs)
+
+
+def _has_pytree_object_arg(schema: torch.FunctionSchema) -> bool:
+    return any(isinstance(arg.type, torch.AnyType) for arg in schema.arguments)
+
+
+def _has_pytree_type_in_args_or_kwargs(args, kwargs) -> bool:
+    return any(
+        not isinstance(x, (list, tuple))
+        and type(x) in torch.utils._pytree.SUPPORTED_NODES
+        for x in itertools.chain(args, kwargs.values())
+    )
```

Comment:
```suggestion
def _has_pytree_type_in_args_or_kwargs(args, kwargs) -> bool:
    return any(
        not torch.utils._pytree.tree_is_leaf(
            x,
            is_leaf=lambda n: isinstance(n, (list, tuple)),
        )
        for x in itertools.chain(args, kwargs.values())
    )
```

You can use `tree_is_leaf` after #113257 get merged into main.